pkg/controller/usersignup/usersignup_controller_test.go

```diff
@@ -1332,6 +1423,212 @@ func TestUserSignupDeactivatedAfterMURCreated(t *testing.T) {
+	key := test.NamespacedName(operatorNamespace, userSignup.Name)
+
+	t.Run("when reactivating the usersignup successfully", func(t *testing.T) {
+		// given
+		InitializeCounter(t, 2)
+		defer counter.Reset()
+		r, req, _ := prepareReconcile(t, userSignup.Name, userSignup, configMap(configuration.UserApprovalPolicyAutomatic), basicNSTemplateTier)
+		createMemberCluster(t, r.client, "member1", ready)
+		defer clearMemberClusters(r.client)
+
+		// when
+		_, err := r.Reconcile(req)
+
+		// then
+		require.NoError(t, err)
+
+		// Lookup the UserSignup
+		err = r.client.Get(context.TODO(), key, userSignup)
+		require.NoError(t, err)
+
+		// Confirm the status has been set to Deactivated
+		test.AssertConditionsMatch(t, userSignup.Status.Conditions,
+			v1alpha1.Condition{
+				Type:   v1alpha1.UserSignupApproved,
+				Status: v1.ConditionTrue,
+				Reason: "ApprovedAutomatically",
+			},
+			v1alpha1.Condition{
+				Type:   v1alpha1.UserSignupComplete,
+				Status: v1.ConditionTrue,
+			},
+			v1alpha1.Condition{
+				Type:   v1alpha1.UserSignupUserDeactivatedNotificationCreated,
+				Status: v1.ConditionFalse,
+				Reason: "NotificationCRNotCreated",
+			})
+		AssertThatCounterHas(t, 3)
+
+		// There should not be a notification created because the user was reactivated
+		notifications := &v1alpha1.NotificationList{}
+		err = r.client.List(context.TODO(), notifications)
+		require.NoError(t, err)
+		require.Len(t, notifications.Items, 0)
+	})
+
+	t.Run("when resetting the usersignup deactivation notification status fails", func(t *testing.T) {
+		// given
+		InitializeCounter(t, 2)
+		defer counter.Reset()
+		r, req, cl := prepareReconcile(t, userSignup.Name, userSignup, configMap(configuration.UserApprovalPolicyAutomatic), basicNSTemplateTier)
+
+		cl.MockStatusUpdate = func(ctx context.Context, obj runtime.Object, opts ...client.UpdateOption) error {
+			switch obj.(type) {
+			case *v1alpha1.UserSignup:
+				return errors.New("failed to update UserSignup status")
+			default:
+				return cl.Client.Update(ctx, obj)
+			}
+		}
+
+		// when
+		_, err := r.Reconcile(req)
+
+		// then
+		require.Error(t, err)
+		require.Equal(t, "failed to update UserSignup status", err.Error())
+
+		// Lookup the UserSignup
+		err = r.client.Get(context.TODO(), key, userSignup)
+		require.NoError(t, err)
+
+		// Confirm the status has been set to Deactivated
+		test.AssertConditionsMatch(t, userSignup.Status.Conditions,
+			v1alpha1.Condition{
+				Type:   v1alpha1.UserSignupApproved,
+				Status: v1.ConditionTrue,
+				Reason: "ApprovedAutomatically",
+			},
+			v1alpha1.Condition{
+				Type:   v1alpha1.UserSignupComplete,
+				Status: v1.ConditionTrue,
+			},
+			v1alpha1.Condition{
+				Type:   v1alpha1.UserSignupUserDeactivatedNotificationCreated,
+				Status: v1.ConditionFalse,
+				Reason: "NotificationCRNotCreated",
+			})
+		AssertThatCounterHas(t, 2)
+
+		// A deactivation notification should not be created because this is the reactivation case
```

Comment:
I'm a bit lost what you are actually testing here?
Is it a test for covering status update case? Then what does the reactivation have with it? I don't even see that the user is reactivated in the test.

Reply:
When a usersignup is deactivated (`Deactivated: true`) then it should get status condition `NotificationCRCreated`. So the test starts with a usersignup has `Deactivated: false` but it has condition `NotificationCRCreated` so this is the scenario where a user was deactivated and was just reactivated. Then the reconcile should update the status to `UserIsActive`. I added comments and initialized the status conditions in the individual test functions to make it a bit more clear. Does it help? Any other suggestions?

Reply:
OK. Makes sense. The comments helped but I think we can make them even clearer :) See my suggestions.